pkg/ipcache/ipcache.go

```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
-		if !allowOverwrite(existingIdentity.Source, identity.Source) {
+	var cidr *net.IPNet
+	var oldIdentity *identity.NumericIdentity
+	callbackListeners := true
+
+	cachedIdentity, found := ipc.ipToIdentityCache[ip]
+	if found {
+		if !allowOverwrite(cachedIdentity.Source, newIdentity.Source) {
 			return false
 		}
 
-		// Skip update if IP is already mapped to the given identity
-		if existingIdentity == identity {
+		// Skip update if IP is already mapped to the given identity.
+		// Ignore the sources of those identities.
+		if cachedIdentity.ID == newIdentity.ID {
 			return false
 		}
-	}
-
-	// An update is treated as a deletion and then an insert.
-	ipc.deleteLocked(IP)
-
-	log.WithFields(logrus.Fields{
-		logfields.IPAddr:   IP,
-		logfields.Identity: identity,
-	}).Debug("Upserting into ipcache layer")
 
-	// Update both maps.
-	ipc.ipToIdentityCache[IP] = identity
-
-	_, found := ipc.identityToIPCache[identity.ID]
-	if !found {
-		ipc.identityToIPCache[identity.ID] = map[string]struct{}{}
+		oldIdentity = &cachedIdentity.ID
 	}
-	ipc.identityToIPCache[identity.ID][IP] = struct{}{}
 
-	// Add a reference for the prefix length if this is a CIDR.
-	if _, cidr, err := net.ParseCIDR(IP); err == nil {
+	// Endpoint IP identities take precedence over CIDR identities, so if the
+	// IP is a full CIDR prefix and there's an existing equivalent endpoint IP,
+	// don't notify the listeners.
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
+		// Add a reference for the prefix length if this is a CIDR.
 		pl, bits := cidr.Mask.Size()
 		switch bits {
 		case net.IPv6len * 8:
 			refPrefixLength(ipc.v6PrefixLengths, pl)
 		case net.IPv4len * 8:
 			refPrefixLength(ipc.v4PrefixLengths, pl)
 		}
+
+		ones, bits := cidr.Mask.Size()
+		if ones == bits {
+			if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+				scopedLog.Info("Upserting CIDR shadowed by endpoint IP")
+				// Skip calling back the listeners, since the endpoint IP has
+				// precedence over the new CIDR.
+				callbackListeners = false
+			}
+		}
+	} else if endpointIP := net.ParseIP(ip); endpointIP != nil { // Endpoint IP.
+		// Convert the endpoint IP into an equivalent full CIDR.
+		bits := net.IPv6len * 8
+		if endpointIP.To4() != nil {
+			bits = net.IPv4len * 8
+		}
+		cidr = &net.IPNet{
+			IP:   endpointIP,
+			Mask: net.CIDRMask(bits, bits),
+		}
+
+		// Check whether the upserted endpoint IP will shadow that CIDR, and
+		// replace its mapping with the listeners if that was the case.
+		if !found {
+			if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+				if cidrIdentity.ID != newIdentity.ID {
+					scopedLog.Info("Upserting endpoint IP shadowing CIDR")
+					oldIdentity = &cidrIdentity.ID
+				} else {
+					// The endpoint IP and the CIDR are associated with the
+					// same identity. Nothing changes for the listeners.
+					callbackListeners = false
```

Comment:
This seems wrong as the endpoint IP based mapping can have a different hostIP that needs to be plumbed.

Reply:
This is not worse than the current behavior. Currently, if we call `Upsert` with the same IP and same ID, but a different host IP, the listeners won't be called back either.

My take is that this is OK since the host IP for an endpoint IP will never change, given how we're allocating endpoint IPs in Cilium. And the initial association between an endpoint IP and its host IP will always be forwarded to the listeners. So it will continue to work as it does now.

If we wanted ipcache to be responsible for mapping endpoint IPs to host IPs, the logic to determine whether to callback the listeners and with what parameters should have been different and more complicated. Notably, we should maintain a map of IP to host IP, because we'd need to keep track of the old host IP so it could be passed to the listeners and so we could determine whether to callback the listeners when the host IP changes.

Reply:
Just to clarify: in which cases does the host IP for an endpoint IP change? I thought that this couldn't happen in Cilium.

Reply:
Is it a bad thing to complain loudly in the logs here if the hostIP doesn't match?

Reply:
Unfortunately, to be able to detect a mismatch I’ll have to maintain a map of IP to host IP. If we’re biting the bullet and manage that mapping anyway, we might as well do it the right way.

Reply:
I don't understand the concern that " the endpoint IP based mapping can have a different hostIP that needs to be plumbed" - when can this be the case? When could such a mismatch occur?

Reply:
The most likely scenario are outdated ipcache entries in the kvstore that then get overwritten by a node reusing a previously used PodCIDR with a changed hostIP. In general I find it highly dangerous to make assumptions in such generic code that is at the foundation of everything. We can exclude this case for now and track it in an issue but we should really turn it around and proof that a hostIP can never change instead of arguing about likeliness of it changing.

Reply:
I updated ipcache to properly handle changing the host IP for an endpoint.